src/util/syscall_sandbox.cpp

```diff
@@ -0,0 +1,830 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif // defined(HAVE_CONFIG_H)
+
+#include <util/syscall_sandbox.h>
+
+#include <array>
+#include <cassert>
+#include <cstdint>
+#include <exception>
+#include <map>
+#include <new>
+#include <set>
+#include <string>
+#include <vector>
+
+#include <logging.h>
+#include <tinyformat.h>
+#include <util/threadnames.h>
+
+#if defined(USE_SYSCALL_SANDBOX)
```

Comment:
I don't think not compiling it is an option since `void SetSyscallSandboxPolicy(SyscallSandboxPolicy syscall_policy)` must be present regardless of `USE_SYSCALL_SANDBOX` (to avoid `ifdef`:s at the various call sites).

If `USE_SYSCALL_SANDBOX` is defined we currently need:

```c++
enum class SyscallSandboxPolicy;
void SetSyscallSandboxPolicy(SyscallSandboxPolicy syscall_policy);
[[nodiscard]] bool SetupSyscallSandbox(bool log_syscall_violation_before_terminating);
void TestDisallowedSandboxCall();
```

If `USE_SYSCALL_SANDBOX` is not defined we only need:

```c++
enum class SyscallSandboxPolicy;
void SetSyscallSandboxPolicy(SyscallSandboxPolicy syscall_policy);
```

In the latter case `SetSyscallSandboxPolicy(SyscallSandboxPolicy syscall_policy)` is a no-op.

I've now simplified the macro use so that in the case of `USE_SYSCALL_SANDBOX` not being defined we get:

```c++
$ cpp -I src/ src/util/syscall_sandbox.cpp | grep -vE '^(#|$)'
enum class SyscallSandboxPolicy {
    INITIALIZATION,
    INITIALIZATION_DNS_SEED,
    INITIALIZATION_LOAD_BLOCKS,
    INITIALIZATION_MAP_PORT,
    INITIALIZATION_TOR_CONTROL,
    MESSAGE_HANDLER,
    NET,
    NET_ADD_CONNECTION,
    NET_HTTP_SERVER,
    NET_HTTP_SERVER_WORKER,
    NET_OPEN_CONNECTION,
    SCHEDULER,
    TX_INDEX,
    VALIDATION_SCRIPT_CHECK,
    SHUTOFF,
};
void SetSyscallSandboxPolicy(SyscallSandboxPolicy syscall_policy);
void SetSyscallSandboxPolicy(SyscallSandboxPolicy syscall_policy)
{
}
```

Another option would be to add `#if defined` at all call sites along the lines of:

```c++
#if defined(USE_SYSCALL_SANDBOX)
SetSyscallSandboxPolicy(SyscallSandboxPolicy::NET);
#endif // defined(USE_SYSCALL_SANDBOX)
```

These are the call sites:

```
git grep SetSyscallSandbox -- ":(exclude)src/util/"
src/bitcoind.cpp:    SetSyscallSandboxPolicy(SyscallSandboxPolicy::SHUTOFF);
src/checkqueue.h:                SetSyscallSandboxPolicy(SyscallSandboxPolicy::VALIDATION_SCRIPT_CHECK);
src/httpserver.cpp:    SetSyscallSandboxPolicy(SyscallSandboxPolicy::NET_HTTP_SERVER);
src/httpserver.cpp:    SetSyscallSandboxPolicy(SyscallSandboxPolicy::NET_HTTP_SERVER_WORKER);
src/index/base.cpp:    SetSyscallSandboxPolicy(SyscallSandboxPolicy::TX_INDEX);
src/mapport.cpp:    SetSyscallSandboxPolicy(SyscallSandboxPolicy::INITIALIZATION_MAP_PORT);
src/net.cpp:    SetSyscallSandboxPolicy(SyscallSandboxPolicy::NET);
src/net.cpp:    SetSyscallSandboxPolicy(SyscallSandboxPolicy::INITIALIZATION_DNS_SEED);
src/net.cpp:    SetSyscallSandboxPolicy(SyscallSandboxPolicy::NET_OPEN_CONNECTION);
src/net.cpp:    SetSyscallSandboxPolicy(SyscallSandboxPolicy::NET_ADD_CONNECTION);
src/net.cpp:    SetSyscallSandboxPolicy(SyscallSandboxPolicy::MESSAGE_HANDLER);
src/node/blockstorage.cpp:    SetSyscallSandboxPolicy(SyscallSandboxPolicy::INITIALIZATION_LOAD_BLOCKS);
src/scheduler.cpp:    SetSyscallSandboxPolicy(SyscallSandboxPolicy::SCHEDULER);
src/torcontrol.cpp:    SetSyscallSandboxPolicy(SyscallSandboxPolicy::INITIALIZATION_TOR_CONTROL);
```

WDYT? :)

Reply:
> WDYT? :)

I like your new solution, minimizing ifdefs in the implementation while also not requiring them at call sites. Conceivably syscall_sandbox could be split up multiple files to avoid using macros, but there wouldn't be a benefit with just one current sandboxing implementation.